docs/how-to-guides/shared-libraries.rst

```diff
@@ -0,0 +1,254 @@
+shared library in a portable manner by using ``install_rpath``:
+
+.. code-block:: meson
+
+    py3.extension_module('_extmodule',
+        '_extmodule.c',
+        link_with: example_lib,
+        install: true,
+        subdir: 'mypkg/subdir',
+        install_rpath: '$ORIGIN'
+    )
+
+The above method will work as advertised on macOS and Linux; ``meson-python`` does
+nothing special for this case. Windows needs some special handling though, due to
+the lack of RPATH support:
+
+.. literalinclude:: ../../tests/packages/sharedlib-in-package/mypkg/__init__.py
+   :start-after: start-literalinclude
+   :end-before: end-literalinclude
+
+If an internal shared library is not only used as part of a Python package, but
+for example also as a regular shared library in a C/C++ project or as a
+standalone library, then the method shown above won't work - the library has to
+be installed to the default ``libdir`` location. In that case, ``meson-python``
+will detect that the library is going to be installed to ``libdir`` - which is
+not a recommended install location for wheels, and not supported by
+``meson-python``. Instead, ``meson-python`` will do the following *on platforms
+other than Windows*:
+
+1. Install the shared library to ``<project-name>.mesonpy.libs`` (i.e., a
+   top-level directory in the wheel, which on install will end up in
+   ``site-packages``).
+2. Rewrite RPATH entries for install targets that depend on the shared library
+   to point to that new install location instead.
+
+This will make the shared library work automatically, with no other action needed
+from the package author. *However*, currently an error is raised for this situation
+on Windows. This is documented also in :ref:`reference-limitations`.
+
+
+External shared libraries
+=========================
+
+External shared libraries are installed somewhere on the build machine, and
+usually detected by a ``dependency()`` or ``compiler.find_library()`` call in a
+``meson.build`` file. When a Python extension module or executable uses the
+dependency, the shared library will be linked against at build time. On
+platforms other than Windows, an RPATH entry is then added to the built
+extension module or executable, which allows the shared library to be loaded at
+runtime.
+
+.. note::
+
+   An RPATH entry alone is not always enough - if the directory that the shared
+   library is located in is not on the loader search path, then the shared
+   library may go missing at runtime. See, e.g., `meson#2121
+   <https://github.com/mesonbuild/meson/issues/2121>`__ and
+   `meson#13046 <https://github.com/mesonbuild/meson/issues/13046>`__ for
+   issues this can cause.
+
+   If this happens, workarounds include adding the directory that the shared
+   library is located in to the search path by amending the ``LD_LIBRARY_PATH``
+   environment variable, or by using a compile flag ``-Wl,-rpath,/path/to/sharedlib/dir``.
```

Comment:
I don't understand this note. Adding an RPATH is equivalent to setting LD_LIBRARY_PATH and the compiler flag is just a way to add to RPATH. Without studying them in detail, what is going on in the linked issues is that RPATH is absolute/relative but the user would like it to be relative/absolute, respectively.

The only way for the library to go missing at load time is when the RPATH is relative and the library install location is changed or the library is not at the location it was when the artifact linking to it was built.

Am I missing something?

Reply:
What is going on is that C/C++/Fortran libraries like `zlib`, `openblas`, etc. expect to be installed into a path that's on the loader search path. As a result, the link line of their `.pc` file doesn't include any RPATH. It will include something like `-lmylib` or `-L/path/to/mylib -lmylib`. This then works fine when the library is indeed in `/usr/lib/` or another standard location. 

However, if you build such a library yourself and put it elsewhere, an RPATH is needed, but not present at all. Meson isn't aware of what directories are or aren't in the loader search path, and hence it's not able to add (or avoid stripping) of RPATHs to such directories at install time. This is particularly confusing because Autotools does know how to do this (xref https://github.com/mesonbuild/meson/issues/13046#issuecomment-2089364514).

It's actually not uncommon for this to bite you in practice if your Python package depends on a shared library, because you'll have it in some custom directory as soon as you want to do any development on it (or put it into its own wheel, like we did with `scipy-openblas32`).

Reply:
I understand the problem you are describing in your latest comment, but I am not able to make the connection from what this PR adds to the documentation and the problem you describe. The documentation added here, almost literally says "An RPATH entry alone is not always enough... if this is the case, add an RPATH" because setting LD_LIBRARY_PATH or the compiler flag just do that.

I now understand what you have in mind, but I think the explanation in the documentation is a bit too terse: I think I have an above average understanding of these issues and still I could not make the connection. I'll try to come up with a concrete suggestion on how to explain this.

Reply:
It's always enough to add your own RPATH by adding -Wl,-rpath=... to your personal LDFLAGS with the same value that you'd add to LD_LIBRARY_PATH, I guess.

The problem is not with rpath, it's with assuming that rpath will as of today be configured out of the box for you.

Reply:
> hence it's not able to add (or avoid stripping) of RPATHs to such directories at install time

Once upon a long time ago meson stripped rpaths unless they were listed in the install_rpath in meson.

These days it fixed that, and now only strips rpaths that *are* listed exclusively in build_rpath (or point to locations inside the build tree).

So your note is not true for versions of meson that meson-python supports.

Reply:
> The problem is not with rpath, it's with assuming that rpath will as of today be configured out of the box for you.

Ah, that's a helpful way of phrasing it I think, thanks.

Reply:
I rewrote this section - shorter now, and hopefully more correct and easier to understand.